File: apps/voice-stream-next/shared/pairing-payload-parser.test.mjs
import { describe, expect, test } from 'bun:test';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const source = readFileSync(path.join(scriptDir, 'pairing-payload-parser.js'), 'utf8');
const module = { exports: {} };
// eslint-disable-next-line no-new-func
new Function('module', 'exports', source)(module, module.exports);
const {
  clientVersionSupported,
  isUpdatePayload,
  parsePairingPayload,
  parseUpdatePayload,
  pairingPayloadExpired,
  webSocketToHttpUrl,
} = module.exports;

describe('pairing-payload-parser', () => {
  test('parses voicestream pairing payloads', () => {
    const payload =
      'voicestream://pair?serverUrl=https%3A%2F%2Fexample.test&deviceId=device-1&token=abc123&displayName=Desktop&deviceType=desktop&minClientVersion=2&expiresAt=2099-01-01T00%3A00%3A00.000Z&pairingSessionId=session-1&apk=https%3A%2F%2Fexample.test%2Fapi%2Fmobile%2Fandroid%2Fapk';

    const config = parsePairingPayload(payload);

    expect(config.serverUrl).toBe('https://example.test');
    expect(config.deviceId).toBe('device-1');
    expect(config.token).toBe('abc123');
    expect(config.deviceName).toBe('Desktop');
    expect(config.minClientVersion).toBe(2);
    expect(config.pairingSessionId).toBe('session-1');
    expect(config.apkUrl).toBe('https://example.test/api/mobile/android/apk');
  });

  test('parses direct websocket URLs with token', () => {
    const config = parsePairingPayload('ws://192.168.1.20:3299/audio?token=abc123');

    expect(config.serverUrl).toBe('http://192.168.1.20:3299');
    expect(config.token).toBe('abc123');
    expect(config.deviceId).toBe('');
  });

  test('parses websocket URLs with device id', () => {
    const config = parsePairingPayload('wss://example.test/audio?token=abc123&deviceId=device-9');

    expect(config.serverUrl).toBe('https://example.test');
    expect(config.deviceId).toBe('device-9');
    expect(config.token).toBe('abc123');
  });

  test('converts websocket URLs to HTTP base URLs', () => {
    expect(webSocketToHttpUrl('wss://example.test:3299/audio?token=abc')).toBe('https://example.test:3299');
    expect(webSocketToHttpUrl('ws://10.0.0.5:3299')).toBe('http://10.0.0.5:3299');
  });

  test('detects update payloads', () => {
    const payload = 'voicestream://update?versionCode=28&apk=https%3A%2F%2Fexample.test%2Fapp.apk';
    expect(isUpdatePayload(payload)).toBe(true);
    expect(parseUpdatePayload(payload)).toEqual({
      versionCode: 28,
      apkUrl: 'https://example.test/app.apk',
    });
  });

  test('rejects invalid payloads', () => {
    expect(() => parsePairingPayload('')).toThrow('Pairing text is empty');
    expect(() => parsePairingPayload('https://example.test/')).toThrow('VoiceStream pairing payload');
    expect(() => parsePairingPayload('wss://example.test/audio')).toThrow('pairing token');
    expect(() => parseUpdatePayload('voicestream://update?apk=https%3A%2F%2Fexample.test%2Fapp.apk')).toThrow('app version');
  });

  test('checks expiry and minimum client version', () => {
    expect(pairingPayloadExpired('2000-01-01T00:00:00.000Z')).toBe(true);
    expect(pairingPayloadExpired('2099-01-01T00:00:00.000Z')).toBe(false);
    expect(clientVersionSupported(1)).toBe(true);
    expect(clientVersionSupported(2)).toBe(false);
  });
});

File: apps/voice-stream-next/android/app/src/test/java/com/huntelkator/voicestreamnext/PairingPayloadParserTest.kt
package com.huntelkator.voicestreamnext

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class PairingPayloadParserTest {
    @Test
    fun parsesVoiceStreamPairingPayload() {
        val payload =
            "voicestream://pair?serverUrl=https%3A%2F%2Fexample.test&deviceId=device-1&token=abc123&displayName=Android&deviceType=android&minClientVersion=2&expiresAt=2099-01-01T00%3A00%3A00.000Z&pairingSessionId=session-1&apk=https%3A%2F%2Fexample.test%2Fapi%2Fmobile%2Fandroid%2Fapk"

        val config = PairingPayloadParser.parse(payload).getOrThrow()

        assertEquals("https://example.test", config.serverUrl)
        assertEquals("device-1", config.deviceId)
        assertEquals("abc123", config.token)
        assertEquals("Android", config.deviceName)
        assertEquals(2L, config.minClientVersion)
        assertEquals("https://example.test/api/mobile/android/apk", config.apkUrl)
    }

    @Test
    fun parsesDirectWebSocketUrlWithToken() {
        val config = PairingPayloadParser.parse("ws://192.168.1.20:3299/audio?token=abc123").getOrThrow()

        assertEquals("http://192.168.1.20:3299", config.serverUrl)
        assertEquals("abc123", config.token)
        assertEquals("", config.deviceId)
    }

    @Test
    fun parsesDirectWebSocketUrlWithDeviceId() {
        val config = PairingPayloadParser.parse("wss://example.test/audio?token=abc123&deviceId=device-9").getOrThrow()

        assertEquals("https://example.test", config.serverUrl)
        assertEquals("device-9", config.deviceId)
        assertEquals("abc123", config.token)
    }

    @Test
    fun convertsWebSocketUrlToHttpBase() {
        assertEquals("https://example.test:3299", PairingPayloadParser.webSocketToHttpUrl("wss://example.test:3299/audio?token=abc"))
        assertEquals("http://10.0.0.5:3299", PairingPayloadParser.webSocketToHttpUrl("ws://10.0.0.5:3299"))
    }

    @Test
    fun parsesUpdatePayload() {
        val apkUrl = "https://example.test/download/app-debug.apk"
        val payload = "voicestream://update?versionCode=28&apk=${encode(apkUrl)}"

        val config = PairingPayloadParser.parseUpdate(payload).getOrThrow()

        assertEquals(28L, config.versionCode)
        assertEquals(apkUrl, config.apkUrl)
        assertTrue(PairingPayloadParser.isUpdatePayload(payload))
    }

    @Test
    fun parsesDesktopAuthPayload() {
        val requestServerUrl = "https://desktop-request.example.test"
        val payload =
            "voicestream://desktop-auth?requestServerUrl=${encode(requestServerUrl)}&requestId=dauth_123&secret=secret-1&deviceToken=token-1&displayName=Desktop&installationId=desktop-install&expiresAt=2099-01-01T00%3A00%3A00.000Z&minClientVersion=3"

        val config = PairingPayloadParser.parseDesktopAuth(payload).getOrThrow()

        assertTrue(PairingPayloadParser.isDesktopAuthPayload(payload))
        assertEquals(requestServerUrl, config.requestServerUrl)
        assertEquals("dauth_123", config.requestId)
        assertEquals("secret-1", config.secret)
        assertEquals("token-1", config.deviceToken)
        assertEquals("Desktop", config.displayName)
        assertEquals("desktop-install", config.installationId)
        assertEquals(3L, config.minClientVersion)
    }

    @Test
    fun parsesDesktopAuthCallbackPayload() {
        val callbackUrl = "http://192.168.1.40:49152/desktop-auth/claim"
        val callbackUrl2 = "http://10.0.0.12:49152/desktop-auth/claim"
        val payload =
            "voicestream://desktop-auth?callbackUrl=${encode(callbackUrl)}&callbackUrls=${encode("[\"$callbackUrl2\",\"$callbackUrl\"]")}&callbackSecret=callback-secret&deviceToken=token-2&displayName=Desktop&installationId=desktop-install"

        val config = PairingPayloadParser.parseDesktopAuth(payload).getOrThrow()

        assertEquals(callbackUrl, config.callbackUrl)
        assertEquals(listOf(callbackUrl2, callbackUrl), config.callbackUrls)
        assertEquals("callback-secret", config.callbackSecret)
        assertEquals("token-2", config.deviceToken)
        assertEquals("Desktop", config.displayName)
    }

    @Test
    fun rejectsUpdatePayloadWithoutVersion() {
        val result = PairingPayloadParser.parseUpdate("voicestream://update?apk=https%3A%2F%2Fexample.test%2Fapp.apk")

        assertTrue(result.isFailure)
    }

    @Test
    fun rejectsPayloadWithoutToken() {
        val result = PairingPayloadParser.parse("wss://example.test/audio")

        assertTrue(result.isFailure)
    }

    @Test
    fun rejectsNonWebSocketPayload() {
        val result = PairingPayloadParser.parse("https://example.test/")

        assertTrue(result.isFailure)
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, StandardCharsets.UTF_8.name())
}
